### ISSUE TITLE
IIS Crash when using SSE

### ISSUE THREAD

**Issue by dlj** [ServiceStack/Issues#616]
This is regarding https://forums.servicestack.net/t/sse-is-bringing-iis-to-its-knees-and-locking-forcing-reboots/5770 as i think i know why. 

I am facing the same problems and i am not running Redis or anything else. Just a lot of clients that all are registering streams. The error i am seeing is the App Poll that is crashing with an AccessViolationException. 
It started when we updated to ServiceStack 5.0.2 and .NET 4.6.x.

After some debugging it seems to be related to WriteEvent and the WriteAsync / FlushAsync code. (Running async in a sync call). Changing the code a little or removing the \n writes greatly lessons the risk of a crash.

I made a small test solution that can crash the IIS every time. It takes between 1-120 seconds depending on how unlucky you are.

[WebApplication1.zip](https://github.com/ServiceStack/Issues/files/2034931/WebApplication1.zip)

**Comment by mythz** [comment]
Thanks for the repro, the async Exceptions weren't should now be properly handled from [this commit](https://github.com/ServiceStack/ServiceStack/commit/f2459e55a1d4a9854f1ce029811accd582d69042).

Now when there's an Exception from writing to a disconnected client, the invalid subscription will be properly clean up. 

Can you try with the latest v5.1.1 that's now [available on MyGet](http://docs.servicestack.net/myget).

**Comment by dlj** [comment]
Sorry but it still happens.

If i remove the "\n" messages, it will continue along fine for hours without any problems.

![devenv_2018-05-25_08-54-59](https://user-images.githubusercontent.com/391714/40530560-911bf31c-5ff9-11e8-8381-7aa1d962f73d.png)

**Comment by mythz** [comment]
> If i remove the "\n" messages, it will continue along fine for hours without any problems.

What do you mean by this?

**Comment by dlj** [comment]
No i am just running the small app i made.
I still takes 15 seconds to crash (and I can reproduce it in production with a couple of user connecting at the same time).

My comment what I tried to change in the code to stop the crash.
The more writes and flushes, the faster it crashes. So removing the newline  that happens at every event, sorta works.

**Comment by mythz** [comment]
Can you please provide the steps to get it to crash, I've been running your App for over an hour and haven't been able to get it crash since. What needs to change for it to crash?

**Comment by dlj** [comment]
Tested on the following setups.
VS 2017 Professional and Enterprice 15.7.2
.Net 4.6.1 and 4.6.2
Physical Windows 10 machines, 1803 and 1709
ServiceStack 5.1.1 from MyGet
IIS Express and IIS 10,

Just running the program will make it crash after 10-20 seconds. Pressing Pause in VS will make it crash even faster.
Is there any other information i can provide to make it easier to debug ? Video, IIS Dump ?

Event Viewer reports the following Error.

Faulting application name: iisexpress.exe, version: 10.0.14358.1000, time stamp: 0x574fc56b
Faulting module name: iiscore.dll, version: 10.0.14358.1000, time stamp: 0x574fc56c
Exception code: 0xc0000005
Fault offset: 0x00026382
Faulting process id: 0x2c70
Faulting application start time: 0x01d3f472f35d7f47
Faulting application path: C:\Program Files (x86)\IIS Express\iisexpress.exe
Faulting module path: C:\Program Files (x86)\IIS Express\iiscore.dll
Report Id: 4c92c8b4-db6a-4199-b8e2-891513027b4f
Faulting package full name: 
Faulting package-relative application ID: 

![image](https://user-images.githubusercontent.com/391714/40567928-eb287fc8-6077-11e8-98c9-d80192d2713f.png)

**Comment by mythz** [comment]
Ok it appears async writes on the current http worker thread is unstable so I've reverted it back to sync writes in [this commit](https://github.com/ServiceStack/ServiceStack/commit/19867385415df739f7aab1ee3d3eb86f6abd6f82).

This change is available from the latest v5.1.1 that's now [available on MyGet](http://docs.servicestack.net/myget).

I'll look at a different approach for async writes and add a flag to switch behavior.

**Comment by dlj** [comment]
I just did another test.

With NotifyChannelOfSubscruptions = False, it crashes in a matter of seconds.

With NotifiChannelOfSubscriptiosns = True, it does not crash, but after 10 minutes this is what my thread count looks like.

![image](https://user-images.githubusercontent.com/391714/40613080-54270dd0-627d-11e8-8fa7-9ef291349419.png)

**Comment by mythz** [comment]
I couldn't repro the crash but I've changed it to avoid blocking locks in [this commit](https://github.com/ServiceStack/ServiceStack/commit/1b3db4b55f6a3a2161a8b885b2a74f9b779fb132).

I've run them with and without `NotifyChannelOfSubscriptions` and with/without a 100ms timer sending notifications and I'm no longer seeing any blocked worker threads in any scenario.

This change is available from v5.1.1 that's now [available on MyGet](http://docs.servicestack.net/myget).

Can you test the latest version (need to clear NuGet cache) and let me know if you're able to repro the crash again.

**Comment by dlj** [comment]
Yes that did the trick. With the newest MyGet packages It does not crash and the there are no stuck Threads.
But i noticed the following things : 

It looks like threads sometimes get stuck in HandleUnlockUnsubscribe and memory there is never released. (See screenshot).
![image](https://user-images.githubusercontent.com/391714/40645464-b39a3de2-6326-11e8-9983-7b7aa906fa14.png)

Stuck Tasks that never recover.
![image](https://user-images.githubusercontent.com/391714/40645559-fda13af8-6326-11e8-9115-8edce21f2c68.png)

**Comment by mythz** [comment]
Can you please list the exact steps to repro the issue.

**Comment by dlj** [comment]
All of them seem to be related to connecting to many times to fast.
In an extreme example, i connect 5 different clients (5 different browsers) that refresh every 100-250 seconds. It will do all the 3 things mentioned above.

This might seem like an extreme example, but we do have 100+ users connecting to the site at the same time. Sometimes more if there are a lot of users and the site is restarted.

**Comment by mythz** [comment]
> 100-250 seconds

Is that seconds or milliseconds?

**Comment by dlj** [comment]
Sorry, yes, that was milliseconds, my bad. 
And as i said, i know that is a lot for a single client. 
But we have 100+ users at a time a time on a site, across multiple sites, that connect in a matter of seconds if we update.

So if there is a memory leak / and or thread lock, it will hit our production environment.

**Comment by mythz** [comment]
In order to resolve the issue I need to be able to repro it, if I can't repro it there's no issue I can fix or be able to verify any changes has, which is why all issues [needs a repro](https://github.com/ServiceStack/Issues#you-can-view-existing-issues-or-report-a-new-issue).

Your reload/re-connection hammer isn't representative of real-world usage since it prevents any heartbeats or unload events from being sent (the browser cancels pending network requests on reloads), at least 1 heartbeat needs to be sent in order to be able to remove expired subscriptions which may be the cause of memory leaks.

I did find one memory leak which is resolved with [this commit](https://github.com/ServiceStack/ServiceStack/commit/d038ad4ad9cdfbc374e17e4f136d4900c6ca1313) which also moves `Dispose()` out of the subscription lock which may resolve a deadlock in `HandleUnsubscription` although I wasn't able to repro any deadlocks after running 5 browser instances of constant new connections for a long time with a timer sending a new msg every 100ms. I'm not even sure the screenshot shows a deadlock since it looks like you only have 2 threads in `HandleUnsubscription`.

Anyway this commit is now available from v5.1.1 that's now on MyGet. If you continue to have issues I'll need a repro. If you're having memory leaks you need to make sure there is at least one connection that's connected long enough to send a heartbeat.

**Comment by mythz** [comment]
I've decided to also call RemoveExpiredSubscriptions after every 1000 connections in case of a rogue DDOS that creates connections without send heartbeats in [this commit](https://github.com/ServiceStack/ServiceStack/commit/a98a50afdd4da138401b059b02352a4b17e49b33).

This change is available in the latest v5.1.1 on MyGet.

**Comment by dlj** [comment]
Thanks. The main problem i had is resolved.
If i can reproduce the other issues i have, without spamming the server will create a new issue.

Many thanks.

**Comment by mythz** [comment]
@dlj glad to hear it 👍
